Cython/Compiler/Symtab.py

```diff
@@ -2430,16 +2442,19 @@ def declare_var(self, name, type, pos,
             cname = name
         entry = self.lookup_here(name)
         if defining and entry is not None:
-            if entry.type.same_as(type):
+            if type.is_cfunction:
+                entry = self.declare(name, cname, type, pos, visibility)
+            elif entry.type.same_as(type):
                 # Fix with_gil vs nogil.
                 entry.type = entry.type.with_with_gil(type.with_gil)
-            elif type.is_cfunction and type.compatible_signature_with(entry.type):
-                entry.type = type
             else:
                 error(pos, "Function signature does not match previous declaration")
         else:
             entry = self.declare(name, cname, type, pos, visibility)
+            if type.is_cfunction and not defining:
+                entry.is_inherited = 1
         entry.is_variable = 1
+        entry.is_cfunction = type.is_cfunction
```

Comment:
I'd only assign this `if type.is_cfunction`. There can be a lot of Entries, so saving instance dict space is a good thing.

Reply:
Thank you for this remark, I wasn't thinking about memory consumption when I made this. Fixed.